### DISCUSSION TITLE
FastAPI/Uvicorn crashes every day around the same time, different APIs, different providers

### DISCUSSION THREAD

**Discussion by JustinGuese** [fastapi/fastapi#6109] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app = FastAPI()


@app.get("/", response_class=HTMLResponse)
async def read_item(request: Request, pac_id: str = None, _oauth2_proxy: str = Cookie(default="")):  
   pass
```

### Description

Three different FastAPI's are crashing every day around the same time (different times per API), running Dockerized on Kubernetes with Uvicorn.

It is not DDOS or anything, as the APIs are using Basic Auth and are behind an nginx proxy which handles authentication.

It is not DB dependent, as one uses MongoDB, one PostgreSQL, one no DB.

Providers are Hetzner, AWS and Advin (Frankfurt, Nürnberg, Amsterdam)

The docker container has a huge CPU and Disk usage before Kubernetes "resets" the container, which causes a downtime of ~5 minutes every day.

I read in some tickets that people experience something similar with uvicorn - is that a problem?
I'm running the app in the docker image with: `CMD ["uvicorn", "main:app", "--host", "0.0.0.0"]` and am using python:3.10-slim

Any recommendations? I would say either a FastAPI problem or Uvicorn

### Operating System

Linux

### Operating System Details

Kubernetes 1.24.9
Docker Image 3.10-slim
Linux Host (Ubuntu Server 22.04)

### FastAPI Version

0.89.0

### Python Version

3.10.9

### Additional Context

![Screenshot 2023-01-24 100520](https://user-images.githubusercontent.com/9588843/214251435-64adb348-f081-48b3-a442-97573d1ae3e3.png)

**Comment by Kludex** [comment]
Do you have the access logs from 9:41 to the time it shut down?

Can you show me the signature of those endpoints?

**Comment by JustinGuese** [comment]
yeah that is the thing, no logs from FastAPI except the typical "GET / 200" (nothing more than usual), and then the Kubernetes message "Container shut down", and the new one spinning up

The signature? Like a model how the architecture looks like?

**Comment by Kludex** [comment]
The signature of the endpoint functions e.g.:

```python
@app.get("/")
def home():
    ...
```

> yeah that is the thing, no logs from FastAPI except the typical "GET / 200" (nothing more than usual), and then the Kubernetes message "Container shut down", and the new one spinning up

Do you have the amount of each request on that period of time?

**Comment by jonatasoli** [comment]
Hello, you set a health check in your api and your pod config?

**Comment by JustinGuese** [comment]
@jonatasoli yes, healthcheck and liveness probe - do you think this might cause it?
```
livenessProbe:
  httpGet:
    path: /new
    port: 8000
  initialDelaySeconds: 5
  periodSeconds: 60
startupProbe:
  httpGet:
    path: /new
    port: 8000
  initialDelaySeconds: 5
  periodSeconds: 10
```

@Kludex yes will post something later, need to anonymize first... It's pretty basic though, so the usual get post stuff

**Comment by JustinGuese** [comment]
and yes i'll implement https://github.com/trallnag/prometheus-fastapi-instrumentator now and then I can send you the exact requests before it crashes tomorrow

**Comment by Kludex** [comment]
Any news?

**Comment by JustinGuese** [comment]
thanks for asking, no at least not on python log leveling... i think somehow fastapi starts using up a lot of ram, which triggers the container kill from kubernetes.... so strange! any ideas how I can get logs that are not shown by the fastapi application? there aren't even requests coming in except the usual healtchecks

**Comment by alexted** [comment]
To understand what's going on with your application, you need to give more information: metrics, logs (add more logging to your application), intercepted exceptions in Sentry, and other debugging information you have.

**Comment by azevakin** [comment]
@JustinGuese what about OOM Killer? requests & limits of resources configured in Kubernetes manifest?

**Comment by yallxe** [comment, marked as answer]
You are running app directly using `uvicorn` command. According to https://www.uvicorn.org/deployment/, when deploying to production you should use `gunicorn -k uvicorn.workers.UvicornWorker`. I also looked up a few examples of docker combined with fastapi, most of them are using gunicorn with uvicorn as a worker class for production.

**Comment by JustinGuese** [comment]
Alright, thanks @yallxe , will try and report back!

**Comment by JustinGuese** [comment]
yes indeed @yallxe's solution of using gunicorn instead of uvicorn solves the problem! So it's a uvicorn bug, weird, but okay! Thanks!!

**Comment by Kludex** [comment]
> yes indeed @yallxe's solution of using gunicorn instead of uvicorn solves the problem! So it's a uvicorn bug, weird, but okay! Thanks!!

What probably happens now is that gunicorn will restart the workers, but the problem should still exist.

**Comment by JustinGuese** [comment]
@Kludex actually I think not. I have tested it in 3 deployments and so far (3 days) no crashes/restarts
But sure maybe gunicorn restarts them so quickly that no traces are seen

**Comment by ghost** [comment]
i am aslo having the same problem and i have been using this command 'gunicorn main:socketio_app --workers 4 --worker-class uvicorn.workers.UvicornWorker --bind 0.0.0.0:8000 --timeout 240' in production, but still the fastapi crashes around 3PM everyday, it restarts itself after few minutes but it is noticeable, i have checked the logs too but didnt find anything suspicious, please suggest what should i do?
